### PR TITLE
T71 keep dates in longwide

### LINKED ISSUE #71: longwide() does not keep dates
The utility function `longwide()` appears to drop dates on input data. It might be helpful to include them in the data returned, to save the hassle of having to re-join data to dates after this step.

The test suite uses `syngrowth`, which does not have dates.

- [ ] make change to function and parameters
- [ ] update function docs in `R/utils.R`
- [ ] update utilities vignette at `vignettes/utilities.Rmd` with examples using new option
- [ ] check quickstart and other vignettes for other references that might require updates
- [ ] rebuild [pkgdown](https://pkgdown.r-lib.org/) docs

### PR DESCRIPTION
**Pull request by molivier-314**
Closes #71 . Allows for longwide() to keep extra columns and include unmatched data with new options. Tests were updated accordingly.

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -136,20 +136,41 @@ recode_sex <- function(input_data,
...
+#' @param inclusion_types Vector indicating which exclusion codes from the
+#' cleaning algorithm should be included in the data, given that include_all is
+#' FALSE. For all options, see growthcleanr::cleangrowth(). Defaults to
+#' c("Include").
+#' @param extra_cols Vector of additional columns to include in the output. If
+#' a column C1 differs on agedays matched height and weight values, then include
+#' separate ht_C1 and wt_C1 columns as well as a match_C1 column that gives
+#' booleans indicating where ht_C1 and wt_C1 are the same. If the agedays
+#' matched height and weight columns are identical, then only include a single
+#' version of C1. Defaults to not keeping any extra columns.
```
Comment (delosh653) [inline R/utils.R:165]: "Defaults to empty vector (not keeping any additional columns)"?

File: R/utils.R [hunk 2]
```diff
@@ -182,101 +203,112 @@ longwide <-
...
+
+    # check that all needed columns are present, if not, throw an error.
+    if(all(c(id, subjid, sex, agedays, param, measurement, gcr_result,
+             extra_cols) %in% colnames(long_df))) {
+      long_df %>%
+        select(id, subjid, sex, agedays, param, measurement, gcr_result,
+               all_of(extra_cols)) -> obs_df
+    } else {
+      # catch error if any variables were not found
+      stop("not all needed columns were present")
```
Comment (delosh653) [inline R/utils.R:221]: "not all specified columns were present"

File: R/utils.R [hunk 3]
```diff
@@ -182,101 +203,112 @@ longwide <-
...
+
+    # join based on subjid, age, and sex, potentially keep unmatched values too
+    wide_df <- merge(height,
+                     weight,
+                     by = c("subjid", "agey", "agem", "agedays", "sex"),
+                     all = keep_unmatched_data) %>%
+      mutate(wt = WEIGHTKG, ht = HEIGHTCM)
+
+
+    for(c in extra_cols) {
```
Comment (delosh653) [inline R/utils.R:290]: Don't use c -- that's a overwriting a function c()

File: R/utils.R [hunk 4]
```diff
@@ -182,101 +203,112 @@ longwide <-
...
+      ht_vals <- wide_df[,paste0(c, ".x")]
+      wt_vals <- wide_df[,paste0(c, ".y")]
+      match_vals <- ht_vals == wt_vals
+      match_vals[is.na(ht_vals) & is.na(wt_vals)] = TRUE
```
Comment (delosh653) [inline R/utils.R:294]: change = to <-

File: tests/testthat/test-utils.R [hunk 5]
```diff
@@ -284,13 +293,234 @@ test_that("longwide works as expected with default values", {
...
+      agedays = sub_syn$agedays,
+      sex = sub_syn$sex,
+      measurement = sub_syn$measurement
+    )
+  )
+
+  # add extra columns to sub_syn; one where all values match, and one where they
+  # don't
+  sub_syn$r1 <- sample(c("A", "B", "C", "D"), size = nrow(sub_syn),
+                       replace = TRUE)
```
Comment (delosh653) [inline tests/testthat/test-utils.R:320]: Add a seed for reproducibility with `set.seed(number)`

File: tests/testthat/test-utils.R [hunk 6]
```diff
@@ -284,13 +293,234 @@ test_that("longwide works as expected with default values", {
...
     expect_equal(wide_syn$wt[w_idx], wt_sub$measurement[wt_idx])
+
+    # check extra columns
+    expect_equal(wide_syn$wt_r1[w_idx], wt_sub$r1[wt_idx])
+    expect_equal(wide_syn$r2[w_idx], wt_sub$r2[wt_idx])
+  }
+})
+
+test_that("longwide works as expected when not dropping unmatched values", {
+  # use synthetic data, running cleaning on a subset
```
Comment (delosh653) [inline tests/testthat/test-utils.R:423]: A lot of this code is copied from above -- could the two tests be merged?

**Review by delosh653 — CHANGES_REQUESTED** [review]

File: R/utils.R (hunk 3 above)
Comment (molivier-314) [reply R/utils.R:290]: Sorry I missed that, change made.

File: tests/testthat/test-utils.R (hunk 6 above)
Comment (molivier-314) [reply tests/testthat/test-utils.R:423]: I went ahead and dropped the spot checks from this test. They are somewhat redundant.

**Review by delosh653 — APPROVED** [review]

**Pull request merged**